pkg/apis/v1/nodepool.go

```diff
@@ -0,0 +1,313 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+
+	"github.com/mitchellh/hashstructure/v2"
+	"github.com/robfig/cron/v3"
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/ptr"
+)
+
+// NodePoolSpec is the top level nodepool specification. Nodepools
+// launch nodes in response to pods that are unschedulable. A single nodepool
+// is capable of managing a diverse set of nodes. Node properties are determined
+// from a combination of nodepool and pod scheduling constraints.
+type NodePoolSpec struct {
+	// Template contains the template of possibilities for the provisioning logic to launch a NodeClaim with.
+	// NodeClaims launched from this NodePool will often be further constrained than the template specifies.
+	// +required
+	Template NodeClaimTemplate `json:"template"`
+	// Disruption contains the parameters that relate to Karpenter's disruption logic
+	// +kubebuilder:default={"consolidationPolicy": "WhenUnderutilized", "expireAfter": "720h"}
+	// +kubebuilder:validation:XValidation:message="consolidateAfter cannot be combined with consolidationPolicy=WhenUnderutilized",rule="has(self.consolidateAfter) ? self.consolidationPolicy != 'WhenUnderutilized' || self.consolidateAfter == 'Never' : true"
+	// +kubebuilder:validation:XValidation:message="consolidateAfter must be specified with consolidationPolicy=WhenEmpty",rule="self.consolidationPolicy == 'WhenEmpty' ? has(self.consolidateAfter) : true"
+	// +optional
+	Disruption Disruption `json:"disruption"`
+	// Limits define a set of bounds for provisioning capacity.
+	// +optional
+	Limits Limits `json:"limits,omitempty"`
+	// Weight is the priority given to the nodepool during scheduling. A higher
+	// numerical weight indicates that this nodepool will be ordered
+	// ahead of other nodepools with lower weights. A nodepool with no weight
+	// will be treated as if it is a nodepool with a weight of 0.
+	// +kubebuilder:validation:Minimum:=1
+	// +kubebuilder:validation:Maximum:=100
+	// +optional
+	Weight *int32 `json:"weight,omitempty"`
+}
+
+type Disruption struct {
+	// ConsolidateAfter is the duration the controller will wait
+	// before attempting to terminate nodes that are underutilized.
+	// Refer to ConsolidationPolicy for how underutilization is considered.
+	// +kubebuilder:validation:Pattern=`^(([0-9]+(s|m|h))+)|(Never)$`
+	// +kubebuilder:validation:Type="string"
+	// +kubebuilder:validation:Schemaless
+	// +optional
```

Comment:
Are we taking our default in here?

Reply:
Talk with @njtran to iron out what this should be.

Reply:
Seems like this is right, at least for now. We should let the rebase after this handle adding this as a default.

Reply:
Yeah that was my current plan